diffcp/cone_program.py

```diff
@@ -309,6 +291,23 @@ def derivative(dA, db, dc, **kwargs):
            NumPy arrays dx, dy, ds, the result of applying the derivative
            to the perturbations.
         """
+        Q = sparse.bmat([
+            [None, A.T, np.expand_dims(c, - 1)],
+            [-A, None, np.expand_dims(b, -1)],
+            [-np.expand_dims(c, -1).T, -np.expand_dims(b, -1).T, None]
+        ])
+
+        D_proj_dual_cone = _diffcp.dprojection(v, cones_parsed, True)
+        if mode == "dense":
+            Q_dense = Q.todense()
+            M = _diffcp.M_dense(Q_dense, cones_parsed, u, v, w)
+            MT = M.T
+
+        pi_z = pi(z, cones)
```

Comment:
The `derivative` callable could conceivably be called more than once for a single solve; in such cases, wouldn't recomputing `Q`, `D_proj_dual_cone` be wasteful (or have I missed something)? I'd prefer pre-computing these quantities and capturing them via lexical closure as before. Otherwise, lazy computation of these quantities would also be fine (i.e., they're computed on the first call to `derivative`, and re-used thereafter). I have the same comment for `adjoint_derivative`.